modules/codegen/src/smithy4s/codegen/ToLine.scala

```diff
@@ -18,92 +18,163 @@ package smithy4s.codegen
+      case Primitive.Int        => TypeReference("Int")
+      case Primitive.Short      => TypeReference("Short")
+      case Primitive.Long       => TypeReference("Long")
+      case Primitive.Float      => TypeReference("Float")
+      case Primitive.Double     => TypeReference("Double")
+      case Primitive.BigDecimal => TypeReference("BigDecimal")
+      case Primitive.BigInteger => TypeReference("BigInt")
+      case Primitive.Uuid       => TypeReference("java.util", "UUID")
+      case Primitive.Document   => TypeReference("smithy4s", "Document")
+      case Primitive.Nothing    => TypeReference("Nothing")
     }
 }
 
 // Models
-case class Line(imports: Set[String], line: String) {
-  def tupled = (imports, line)
-  def suffix(suffix: Line) = modify(s => s"$s $suffix")
-  def addImport(imp: String) = copy(imports = imports + imp)
-  def modify(f: String => String) = Line(imports, f(line))
-  def nonEmpty = line.nonEmpty
-  def toLines = Lines(imports, List(line))
-
-  def args(l: LinesWithValue*): Lines = if (l.exists(_.render.lines.nonEmpty)) {
-    val openBlock = if (line.nonEmpty) line + "(" else ""
-    Lines(imports, List(openBlock)) ++ indent(
-      l.toList.foldMap(_.render).mapLines(_ + ",")
-    ) ++ Lines(")")
-  } else Lines(line + "()").addImports(imports)
 
+sealed trait LineSegment { self =>
+  def toLine: Line = Line(Chain.one(self))
+}
+object LineSegment {
+  case class Import(value: String) extends LineSegment
+  case class Hardcoded(value: String) extends LineSegment
+  case class TypeDefinition(name: String) extends LineSegment
+  case class TypeReference(pkg: List[String], name: String)
+      extends LineSegment { self =>
+    def asValue: String = s"${(pkg :+ name).mkString(".")}"
+  }
+  object TypeReference {
+    def apply(pkg: String, name: String): Line =
+      TypeReference(pkg.split("\\.").toList, name).toLine
+    def apply(fqn: String): Line = {
+      val parts = fqn.split("\\.").toList
+      TypeReference(parts.dropRight(1), parts.last).toLine
+    }
+
+  }
+
+  implicit val hardcodedShow = Show.show[Hardcoded](_.value)
+  implicit val importShow = Show.show[Import](_.value)
+  implicit val typeDefShow: Show[TypeDefinition] = Show.show { td =>
+    s"${td.name}"
+  }
+  implicit val typeRefShow: Show[TypeReference] = Show.show { tr =>
+    s"${(tr.pkg :+ tr.name.split("\\.")(0)).mkString(".")}"
+  }
+  implicit val lineSegmentShow: Show[LineSegment] = Show.show {
+    case Import(value)      => value
+    case Hardcoded(value)   => value
+    case td: TypeDefinition => td.show
+    case tr: TypeReference  => tr.show
+  }
+  implicit val lineShow: Show[Line] =
+    Show.show(line => line.segments.toList.map(_.show).mkString(""))
+
+  implicit def chainShow[A: Show]: Show[Chain[A]] = Show.show { chain =>
+    chain.toList.map(_.show).mkString
+  }
+}
+case class Line(segments: Chain[LineSegment]) {
+  self =>
+  def :++(other: Line): Line = Line(self.segments ++ other.segments)
```

Comment:
I'm personally finding these operators hard to read. Not your fault really, because you've copied the ones from the stdlib (which I also find hard to read, and much prefer the english-ones). 

Because in this usecase we're not dealing with generic collections, I think it's better to have a unique operator `+` that would be overloaded. This would echo the UX of dealing with strings in Scala (see screenshot below) 

![Screenshot 2022-08-12 at 09 53 30](https://user-images.githubusercontent.com/4439335/184310202-cec79261-ee03-47b7-926e-102e7a9b6129.png)

This operator would be present as methods on both `Line` and `LineSegment`,  one version taking a `Line` parameter, another version taking a `LineSegment`.